<div style="border:solid green 2px; padding: 20px">
<b>Эмин, привет!</b>

Меня зовут Евгений Дащенко, и я буду проверять твой проект. Предлагаю общаться на «ты» 😊 Но если это не удобно - дай знать, и мы перейдем на "вы". 

Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе стать аналитиком данных. Ты уже проделал большую работу над проектом, но давай сделаем его еще лучше. Ниже ты найдешь мои комментарии - пожалуйста, не перемещай, не изменяй и не удаляй их. Я буду использовать цветовую разметку:

<div class="alert alert-danger">
<b>Критическая ошибка:</b> Так выделены самые важные замечания. Без их отработки проект не будет принят. </div>

<div class="alert alert-warning">
<b>Хорошо, но можно и лучше:</b> Так выделены небольшие замечания. Одна-два таких замечания в проекте допустимы, но если их будет больше - тебе будет необходимо внести исправления. Это как тестовое задание при приеме на работу: очень много мелких ошибок могут стать причиной отказа кандидату. 

</div>

<div class="alert alert-success">
<b>Все отлично:</b> Так я выделяю все остальные комментарии.</div>

Давай работать над проектом в диалоге: если ты что-то меняешь в проекте или отвечаешь на мои комменатри — пиши об этом. Мне будет легче отследить изменения, если ты выделишь свои комментарии:
<div class="alert alert-info"> <b>Комментарий студента:</b> Например, вот так.</div>

Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Здорово, когда есть вступление в сам проект, каждый сможет быстрее разобраться в концепции / деталях работы. 

</div>

## Шаг 1. Откройте файл с данными и изучите общую информацию

In [1]:
import pandas as pd
df = pd.read_csv('/datasets/data.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [2]:
# Дополнительно посмотрим на данные таблицы.
df.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Вывод**

В таблице 21525 строк.  
В столбцах days_employed и total_income есть пропуски. Пропуски более, чем у 10% данных. Удаление этих данных может повлиять на итоги исследования. Поэтому пропуски будут заполняться вручную. 


Столбцы с категориальными переменными: education, education_id, family_status, family_status_id, gender, income_type, debt, purpose.  
Столбцы с количественными переменными: children, days_employed, dob_years, total_income.

Пропуски данных в столбцах days_employed (общий трудовой стаж в днях) и в total_income (ежемесячный доход). Так как значения переменных этих столбцов количественные, пропуски будут заполняться характерными значениями.  

Просмотр данных таблицы позволил увидеть ошибки заполнения данных. Столбец days_employed имеет отрицательные значения.  
По первым 20 строкам можно предположить, что отрицательные значения могут быть связаны со столбцом income_type. Значения отрицательные у всех категорий заемщиков, кроме пенсионеров. Возможно, знак минус имеет смыслоразличительное значение в данном случае.   

Так как  для решения поставленной задачи стаж работы заемщиков не требуется, указанные выше гипотезы в данный момент проверяться не будут. Но заполним и эти пропуски данными для будущих исследований.  

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Достаточно детально изучили наши первичные данные, давай теперь их обработаем. 
</div>

## Шаг 2. Предобработка данных

### Обработка пропусков

In [3]:
# Проверим строки с пропусками в столбце days_employed
df[df['days_employed'].isna()]

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

In [4]:
# После проверки пустых значений в столбце days_employed появилось предположение, что пропуски в столбцах total_income
#и days_employed будут у одних и тех же строк.
# Проверим это:
df[(df['days_employed'].isna()) & (df['total_income'].isna())]

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

In [5]:
# Предположение оказалось верным. Всего, по полученным в самом начале данным, пропусков 2174, в полученном массиве их также 2174.

# Заполним пропуски в столбце трудового стажа (столбец days_employed). 
# Медиана будет использоваться для того, чтобы избежать неточностей метода расчета среднего арифметического
#в ситуациях, когда в данных есть слишком маленькие или слишком большие значения, которые могут повлиять на результат.

# Создаем массив с индексами строк с пустыми значениями для замены значений циклом 
df_nan_rows = df[df['days_employed'].isna()].index

# Создаем массив с уникальными значениями возраста для заполнения пропущенных значений возраста циклом.
unique_dob_years = df['dob_years'].unique()

# Заполняем пропуски (решил делать циклом, а через .fillna(), т.к. при замене пропусков в столбце с доходом 
#цикл будет нагладнее, хоть и занимать больше времени для обработки):
for row in df_nan_rows:
    for year in unique_dob_years:
        if df.loc[row, 'dob_years'] == year:
            df.loc[row, 'days_employed'] = df[df['dob_years'] == year]['days_employed'].median()

# Проверяем массив на наличие пустых значений в столбце days_employed:
df[df['days_employed'].isna()]

# Пустых значений не осталось. 

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

In [6]:
# Недостающую информацию о доходах будем также заполнять медианными значениями.
# Расчет медианных значений для определения дохода каждого конкретного заемщика будем брать исходя из количества детей, 
#образования и типа занятости.

# Делаем массивы для уникальных значений кол-ва детей, образования и типа занятости:
df_children_unique = df['children'].unique()
df_education_unique = df['education_id'].unique()
df_income_type_unique = df['income_type'].unique()

# Циклом осуществляем перебор значений:
for row in df_nan_rows:
    for children_unique in df_children_unique:
        for education_unique in df_education_unique: 
            for income_type_unique in df_income_type_unique: 
                if (df.loc[row, 'children'] == children_unique) & (df.loc[row, 'education_id'] == education_unique) & (df.loc[row, 'income_type'] == income_type_unique):
                    df.loc[row, 'total_income'] = df[(df['children'] == children_unique) & (df['education_id'] == education_unique) & (df['income_type'] == income_type_unique)]['total_income'].median()

# Проверим наличие нулевых значений через .info():
df.info()

# Все пропуски заполнены.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Интересная идея заполнить пропуски исходя из групп в данных 👍, таким образом мы точечно повышаем уровень дисперсии.
    
---
    
Единственное, я все таки рекомендую избегать циклы в обработке данных, т.к. они могут занимать большое количество времени. Альтернативой может быть отдельно созданная функция, котороая потом применяется с помощью метода `apply`. 
    
</div>

**Вывод**

Отсутствие данных в столбцах days_employed и total_income связано между собой. Данные о ежемесячном доходе отсутствуют по тем строкам, по которым отсутствует общий трудовой стаж.  
Возможными причинами отсутствия данных могут быть: ошибки при вводе данных, отсутствие официального стажа работы у конкретных заемщиков (и, соответственно, официального ежемесячного дохода).  
Также предполагаю, что если ведется дополнительная проверка стажа и дохода со стороны кредитного отдела банка, то отсутствие данных может являться следствием того, что предоставленные заемщиками данные не прошли проверку.  

Данные о стаже были заполнены медианными значениями для каждого возраста отдельно.  
Данные о ежемесячном доходе были заполнены медианными значениями на основании имеющейся статистики в таблице с учетом количества детей у заемщиков, уровня образования и типа занятости.


### Замена типа данных

In [7]:
# Заменим тип данных у столбцов days_employed и total_income с float на int (данные этих столбцов
#не требуют детализации до долей, целого остатка будет достаточно для проведения анализа).
# Для замены типа данных используем метод .astype(), так как с его помощью можно перевести тип данных столбцов 
#в требуемый нам  int.
df['days_employed'] = df['days_employed'].astype('int')
df['total_income'] = df['total_income'].astype('int')

# Проверим DataFrame:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


**Вывод**

Мы заменили тип данных у столбцов days_employed и total_income с float на int, т.к. данные этих столбцов
не требуют детализации до долей, целого остатка будет достаточно для проведения анализа.  
Столбцы education, family_status, gender, income_type и purpose хранят текстовые данные. Для них тип данных object подходит.  
Остальные стобцы имеют тип данных int, что также нам подходит.

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

В том числе метод `fillna` и `astype` можно применять к нескольким столбцам одновременно.
    
Если столбцов более 2–ух, то рекомендую применять цикл `for`.
    
Пример: 
    
`df.astype({ "col1": "float64", "col2": "int64"})`
    
</div>

### Обработка дубликатов

In [8]:
# Проверим кол-во полных дубликатов в таблице:
df.duplicated().sum()

# Всего 54 задублированные строки. 
# Т.к. это меньше половины доли процента, удаление этих данных не повлияет на итоги исследования. 

54

In [9]:
# Удаляем строки-дубли:
df = df.drop_duplicates().reset_index(drop = True)

# Перепроверяем кол-во полных дубликатов в таблице:
df.duplicated().sum()

# Строк-дублей в таблице больше нет.

0

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Хорошо, что удалили дубликаты, они нам ни к чему.
    
</div>

In [10]:
# Проверим дубли в столбце education:
df['education'].value_counts()

# В столбце много неявных дублей, которые можно устранить посредством приведения всех букв к единому регистру.

среднее                13705
высшее                  4710
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   273
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

In [11]:
df['education'] = df['education'].str.lower()
df['education'].value_counts()
# Дубликаты столбца education заменены.

среднее                15188
высшее                  5251
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

In [12]:
# Проверим дубли в столбце family_status:
df['family_status'].value_counts()

# В этом столбце дубликатов нет.

женат / замужем          12344
гражданский брак          4163
Не женат / не замужем     2810
в разводе                 1195
вдовец / вдова             959
Name: family_status, dtype: int64

In [13]:
# Проверим дубли в столбце income_type:
df['income_type'].value_counts()

# Дубликатов в этом стоблце нет. Можно обратить внимание на статусы "сотрудник", "компаньон" и "предприниматель". 
# Но без дополнительной консультации с заказчиком для самостоятельного принятия решения не хватает информации. 

сотрудник          11091
компаньон           5080
пенсионер           3837
госслужащий         1457
предприниматель        2
безработный            2
студент                1
в декрете              1
Name: income_type, dtype: int64

**Вывод**

В таблице было 54 задублированные строки. Мы удалили строки-дубли и обновили индексацию строк.  
Дубли строк могли появиться в случае ошибки ввода, или в случае повторного обращение того же заемщика в кредитный отдел банка.  

Далее мы проверили  неявные дубликаты в столбцах education, family_status и income_type. Устранили дубликаты в столбце education, связанные с разницей регистров букв. Эти ошибки связаны, скорее всего, с тем, что данное поле заполняется сотрудниками вручную, и нет маски ввода/ограничений по вводу данных в поле.

<div class="alert alert-warning">
<b>Комментарий ревьюера:</b>

После того как мы произвели столбец полностью в нижний регистр `df['education']`, у нас появились еще дубликаты. Дело в том, что питон воспринимает верхний и нижний регистр как разные символы, поэтому рекомендовал бы еще раз проверять дубли в данных.
    
Будем иметь в виду на будущее этот нюанс.
    

    
</div>

In [30]:
df.duplicated().sum()

17

### Лемматизация

In [14]:
# Импортируем библиотеку pymystem3 для лемматизации данных о целях получения кредита.
from pymystem3 import Mystem
m = Mystem() 

# Лемматизируем данные, добавляем их в отдельную таблицу tags:
tags = []
for i in range(len(df['purpose'])):
    tags.append(m.lemmatize(df.loc[i, 'purpose']))

# Добавляем леммитизированные данные в новый столбец purpose_tags таблицы tags:    
df['purpose_tags'] = pd.Series(tags)

# Проверяем таблицу с новым столбцом
df

children  days_employed  dob_years education  education_id  \
0             1          -8437         42    высшее             0   
1             1          -4024         36   среднее             1   
2             0          -5623         33   среднее             1   
3             3          -4124         32   среднее             1   
4             0         340266         53   среднее             1   
...         ...            ...        ...       ...           ...   
21466         1          -4529         43   среднее             1   
21467         0         343937         67   среднее             1   
21468         1          -2113         38   среднее             1   
21469         3          -3112         38   среднее             1   
21470         2          -1984         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21466  гражданский брак                 1      F   компаньон     0   
21467   женат / замужем                 0      F   пенсионер     0   
21468  гражданский брак                 1      M   сотрудник     1   
21469   женат / замужем                 0      M   сотрудник     1   
21470   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose  \
0            253875                 покупка жилья   
1            112080       приобретение автомобиля   
2            145885                 покупка жилья   
3            267628    дополнительное образование   
4            158616               сыграть свадьбу   
...             ...                           ...   
21466        224791             операции с жильем   
21467        155999          сделка с автомобилем   
21468         89672                  недвижимость   
21469        244093  на покупку своего автомобиля   
21470         82047         на покупку автомобиля   

                                       purpose_tags  
0                           [покупка,  , жилье, \n]  
1                 [приобретение,  , автомобиль, \n]  
2                           [покупка,  , жилье, \n]  
3              [дополнительный,  , образование, \n]  
4                         [сыграть,  , свадьба, \n]  
...                                             ...  
21466                [операция,  , с,  , жилье, \n]  
21467             [сделка,  , с,  , автомобиль, \n]  
21468                            [недвижимость, \n]  
21469  [на,  , покупка,  , свой,  , автомобиль, \n]  
21470           [на,  , покупка,  , автомобиль, \n]  

[21471 rows x 13 columns]

**Вывод**

Мы лемматизировали (привели все слова столбца purpose к их леммам - к их словарным формам) информацию о целях получения кредита для упрощения последующего анализа влияния цели кредита на своевременный возврат кредита в срок. 

### Категоризация данных

In [15]:
# Вынесем уровень образования и семейный статус в отдельные словари для удобства дальнейшего обращения к данным таблиц.
# Сделаем словарь для уровня образования:
education_dict = df[['education_id', 'education']]
education_dict = education_dict.drop_duplicates().reset_index(drop = True)
education_dict

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

In [16]:
# Делаем аналогичный словарь для семейного статуса:
family_status_dict = df[['family_status_id', 'family_status']]
family_status_dict = family_status_dict.drop_duplicates().reset_index(drop = True)
family_status_dict

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  Не женат / не замужем

In [17]:
# Категоризируем данные столбца purpose_tags для дальнейшего анализа.
# Для этого сначала посмотрим, какие уникальные значения встречаются в этом столбце:

df['purpose_tags'].value_counts()

[автомобиль, \n]                                          972
[свадьба, \n]                                             793
[на,  , проведение,  , свадьба, \n]                       773
[сыграть,  , свадьба, \n]                                 769
[операция,  , с,  , недвижимость, \n]                     675
[покупка,  , коммерческий,  , недвижимость, \n]           662
[операция,  , с,  , жилье, \n]                            652
[покупка,  , жилье,  , для,  , сдача, \n]                 652
[операция,  , с,  , коммерческий,  , недвижимость, \n]    650
[покупка,  , жилье, \n]                                   646
[жилье, \n]                                               646
[покупка,  , жилье,  , для,  , семья, \n]                 638
[строительство,  , собственный,  , недвижимость, \n]      635
[недвижимость, \n]                                        633
[операция,  , со,  , свой,  , недвижимость, \n]           627
[строительство,  , жилой,  , недвижимость, \n]            625
[покупка

In [18]:
# Все цели можно поделить на следующие категории: 1) автомобиль; 2) свадьба; 3) недвижимость/жилье; 4) образование.
# Заменим данные типа Series в столбце purpose_tags на выделенные выше категории. 
# Для этого напишем функцию, которая пройдет по всем строкам массива, и возвратит нужную категорию, 
#если она присутствует в строке. Добавим в функцию условие try/except для избежания ошибок при отработке кода:

def categorize(data):
    for i in range(len(data['purpose_tags'])):
        try:
            if 'автомобиль' in data.loc[i, 'purpose_tags']:
                data.loc[i, 'purpose_tags'] = 'автомобиль'
            elif 'свадьба' in data.loc[i, 'purpose_tags']:
                data.loc[i, 'purpose_tags'] = 'свадьба'
            elif 'недвижимость' in data.loc[i, 'purpose_tags']:
                data.loc[i, 'purpose_tags'] = 'недвижимость'
            elif 'жилье' in data.loc[i, 'purpose_tags']:
                data.loc[i, 'purpose_tags'] = 'недвижимость'
            elif 'образование' in data.loc[i, 'purpose_tags']:
                data.loc[i, 'purpose_tags'] = 'образование'
        except:
            data.loc[i, 'purpose_tags'] = 'error'

# Так как жилье и недвижимость синонимичные понятия, в обоих случаях будем присваивать категорию 'недвижимость'.
# Применим функцию к таблице, заменим значение столбца purpose_tags на категории:

categorize(df)


<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Категории кредитов опредилили верно 👍
    
</div>

In [19]:
# Переименуем столбец purpose_tags в category и проверим таблицу на предмет наличие категории error:

df.rename(columns={'purpose_tags': 'category'}, inplace=True)
df['category'].value_counts()

недвижимость    10814
автомобиль       4308
образование      4014
свадьба          2335
Name: category, dtype: int64

**Вывод**

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Отличный и безопасный способ заменить названия в столбцах 👍
    
</div>

Были созданы словари уровня образования и семейного статуса заемщиков для дальнейшего упрощения работы с данными таблицы.  

Выделенные в рамках лемматизации словарные формы слов были категоризированы на четыре группы.
Основные причины, по которым заемщики брали кредит: 1) автомобиль; 2) свадьба; 3) недвижимость; 4) образование. 
Эти категории были добавлены в таблицу в столбец 'category'.

## Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [20]:
# Проверим количество уникальных значений в столбце children: 
df['children'].value_counts()

 0     14107
 1      4809
 2      2052
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [21]:
# В данных есть артефакты: -1 ребенок и 20 детей. Всего этих значений 123. Это чуть больше 0,57% от общего количества данных. 
# Можно предположить, что "-1" - это написанное с ошибкой "1", а "20" - это "2". У нас нет возможности 
#проконсультироваться с заказчиком для определения точной причины этих ошибок. 
# Так как количество этих данных мал в масштабе массива данных, было бы целесообразно удалить эти данные.
# Но в связи с тем, что нам потребуется рассчитать не только зависимость между наличием детей и возвратом кредита в срок, 
#оставим эти данные, скорректировав данные о количестве детей на "1" и "2".

df['children'] = df['children'].replace(-1, 1)
df['children'] = df['children'].replace(20, 2)

# Проверяем обновленные данные по столбцу children:
df['children'].value_counts()

0    14107
1     4856
2     2128
3      330
4       41
5        9
Name: children, dtype: int64

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

А если бы мы посмотрели на долю этих значений, то смогли бы принять решение в сторону избавления от этих данных ? 
    
</div>

In [22]:
# Проверим количество долгов в разрезе количества детей:
df.pivot_table(index=['children'], values='debt', aggfunc='sum')
print(df.pivot_table(index=['children'], values='debt', aggfunc='sum')) 

          debt
children      
0         1063
1          445
2          202
3           27
4            4
5            0


In [23]:
# Выведем процент заемщиков с наличинем задолженностей по возврату кретитов в разрезе количества детей, чтобы ответить
#на поставленный вопрос. 
# Выведем таблицу через display(), отсортируем ее по возрастанию доли задолженности,
#и перемиенуем столбец с итогами расчетов для большей наглядности:
df_child_pivot = df.pivot_table(index=['children'], values='debt', aggfunc='mean').sort_values('debt')
df_child_pivot.rename(columns={'debt': 'Доля заемщиков с задолженностью'}, inplace=True)
display(df_child_pivot) 

Доля заемщиков с задолженностью
children                                 
5                                0.000000
0                                0.075353
3                                0.081818
1                                0.091639
2                                0.094925
4                                0.097561

**Вывод**

Согласно предоставленным заказчиком данным, определенная зависимость между наличием детей и вовзвратом кредита в срок наблюдается, хоть корреляция и не сильно выраженная.  
Люди, не имеющие детей, чаще возвращают кредит в срок, нежели чем те, кто имеет от одного до четырех детей (ок. 7,5% заемщиков без детей имели ранее задолженности по кредитам).  
Среди тех заемщиков, у кого от одного до четырех детей, наименьшее количество задолженностей по возвратам кредитов имели те, у кого три ребенка (ок. 8,2% заемщиков имели задолженности). Далее идут заемщики с одним ребенком (~9,5%). Наибольшее количество задолженностей по кредитам имели заемщики с четыремя детьми - 9,7% заемщиков.  
У людей с 5 детьми не имелось задолженностей по возврату кредитов вообще. Однако данная выборка нерепрезентативна - в таблице предоставлена информация о 9 заемщиках с пятью детьми; данных недостаточно, чтобы экстраполировать выводы для построения модели кредитного скроллинга.

Результаты полученного среза могут быть связаны с тем, что наличие детей налагает на человека больше обязательств (в т.ч. финансовых), а также предполагает больше непредсказуемых трат.  
Неожиданно меньше задолженностей по кредитам у людей с тремя детьми. Возможно, это связано с тем, что люди, имеющие трех детей, более взвешенно принимали решении о рождении третьего ребенка (имея понимание о расходах и обязательствах). Допускаю, своего рода, "эффект масштаба".

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Верно 👍
    
</div>

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [24]:
print(df['family_status'].value_counts())

женат / замужем          12344
гражданский брак          4163
Не женат / не замужем     2810
в разводе                 1195
вдовец / вдова             959
Name: family_status, dtype: int64


In [25]:
df_fam_stat_pivot = df.pivot_table(index=['family_status'], values='debt', aggfunc='mean').sort_values('debt')
df_fam_stat_pivot.rename(columns={'debt': 'Доля заемщиков с задолженностью'}, inplace=True)
display(df_fam_stat_pivot) 

Доля заемщиков с задолженностью
family_status                                         
вдовец / вдова                                0.065693
в разводе                                     0.071130
женат / замужем                               0.075421
гражданский брак                              0.093202
Не женат / не замужем                         0.097509

**Вывод**

Зависимость между семейным положением и возвратом кредита в срок также прослеживается. В данном случае более явная, нежели чем в первом случае.  
Чаще всего в срок возвращали кредит те заемщики, которые утратили своего партнера (ок. 6,6% заемщиков имели просроченные кредиты). Больше просроченных задолженностей имели люди в разводе и в официальном браке (7,1% и 7,5% просроченных кредитов).  
Наибольшее количество задолженностей по кредитам имели люди в граджанском браке и холостые люди (9,3% и 9,7%).  

Наличие официальных отношений (даже бывших) может говорить о том, что человек более ответственно относится к возврату кредита в срок.

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

В точку 👍
    
Обрати внимание, что наши действия повторяются с точки зрения расчетов.
    
Это говорит о том, что мы можем создать функцию для постоянных расчетов. Тем самым мы сможем оптимизировать свой код 😊
    
</div>

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [26]:
# Для корректного анализа зависимости между уровнем дохода и возвратом кредита в срок потребуется 
#категоризировать уровни дохода.
# Разобьем данные на следующие категории: 1) до 100000; 2) от 100000 до 150000; 3) от 150000 до 200000; 4) более 200000. 
# Добавим категории в отдельный столбец income_category

for i in range(len(df['total_income'])):
    if df.loc[i, 'total_income'] < 100000:
        df.loc[i, 'income_category'] = 'до 100000'
    elif 100000 <= df.loc[i, 'total_income'] < 150000:
        df.loc[i, 'income_category'] = 'от 100000 до 150000'
    elif 150000 <= df.loc[i, 'total_income'] < 200000:
        df.loc[i, 'income_category'] = 'от 150000 до 200000'
    else: df.loc[i, 'income_category'] = 'более 200000'
  
        
# Категории были выбраны в соответствии с количеством данных в таблице.

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Хороший способ категоризовать данные по уровню дохода 👍
    
Думаю теперь разработать любую категорию в наших данных не составит труда.

</div>

In [27]:
df['income_category'].value_counts()  

от 100000 до 150000    6991
более 200000           5213
от 150000 до 200000    4804
до 100000              4463
Name: income_category, dtype: int64

In [28]:
df_income_pivot = df.pivot_table(index=['income_category'], values='debt', aggfunc='mean').sort_values('debt')
df_income_pivot.rename(columns={'debt': 'Доля заемщиков с задолженностью'}, inplace=True)
display(df_income_pivot) 

Доля заемщиков с задолженностью
income_category                                     
более 200000                                0.069825
до 100000                                   0.079319
от 150000 до 200000                         0.085137
от 100000 до 150000                         0.087827

**Вывод**

Все заемщики были разбиты на 4 категории в зависимости от уровня дохода: 1) до 100000; 2) от 100000 до 150000; 3) от 150000 до 200000; 4) более 200000.
Зависимость между возвратом кредитов в срок и доходом таккже прослеживается. 
Меньше всего задолженностей по кредитам у людей с достатком более 200000 руб. (почти 7%). Далее идет категория с доходом до 100000 руб. (7,9% заемщиков с долгами по кредитам).  
У остальных двух категорий 8,5% и почти 8,8% заемщиков имели задолженность за кредиты. Наибольшее количество задолженностей у категории людей с доходом от 100000 до 150000 руб.  

Полагаю, что у людей с доходом более 200000 руб. есть больше возможностей для своевременного возврата кредита.  
Люди с доходом до 100000 руб. чуть более бережно обходятся со своими финансками, чем остальные заемщики (с доходом до 200000 руб.).

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Верно 👍
    
В том числе можно было посчитать с помощью альтернативного метода группировки `groupby`, будем иметь в виду на будущее.
    
</div>

- Как разные цели кредита влияют на его возврат в срок?

In [29]:
df_purpose_pivot = df.pivot_table(index=['category'], values='debt', aggfunc='mean').sort_values('debt')
df_purpose_pivot.rename(columns={'debt': 'Доля заемщиков с задолженностью'}, inplace=True)
display(df_purpose_pivot) 

Доля заемщиков с задолженностью
category                                     
недвижимость                         0.072314
свадьба                              0.079657
образование                          0.092177
автомобиль                           0.093547

**Вывод**

Цели кредита также влияют на своевременный возврат кредита в срок. 
Те, кто берет кредит на недвижимость и на свадьбу, чаще не имеют задолженностей по кредитам, нежели чем те, кто берет кредит на образование или для покупки автомобиля. 
Наименьшая доля задолженностей по кредитам среди тех, кто берет кредит на покупку недвижимости (7,2%). Кредит на покупку недвижимости важный и отвественный шаг, поэтому человек, который планирует брать кредит на эти нужды, заранее заботится о своей кредитной истории.
Следующая категория - заемщики, которые берут кредиты на свадьбу (ок. 8% заемщиков имеют задолженности по кредитам).
Наибольшее количество задолженностей по кредитам - кредиты, получаемые на образование и на приобретение автомобия (9,2% и 9,3%).  

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

В точку 👍
    
В том числе можно было бы построить графики, например с помощью метода `hist` 
    
Как раз способы визуализации, мы уже будем проходить в следующем проекте 😊
    
</div>

## Шаг 4. Общий вывод

При построении модели кредитного скоринга желательно учесть наличие детей у заемщика, его семейное положение, уровень дохода и цели кредита. Разница в доле заемщиков, которые не вернут кредит в срок, между категориями каждого среза может составлять до 3%.  
Идеальный портрет заемщика: вдова без детей с доходом от 200000 рублей, которая приобретает в кредит недвижимость. 

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Выводы понятны и логичны, а самое главное подкреплены выявленными фактами.
       
В целом сам проект выполнен на хорошем уровне! На протяжении всей работы чувствуется глубина проработки задачи и это безусловный плюс для нас.
    
Хочется отметить, что мы узнали достаточно много нового в этом проекте и с уверенностью можем двигаться дальше. На что рекомендую обращать внимание сразу с самого старта так сказать, а именно на повторяемость кода. Если наши действия начинают повторяться более 2-х раз, то имеет смысл задуматься над: 
    
- внедрением цикла 
- внедрением функции
- внедрением комбинации цикл + функция
    
Такие вещи могут сразу не бросаться в глаза, но постепенно мы придём к такому подходу 😊
    
Поздравляю с успешным завершением проекта 😊👍
    
И желаю успехов в новых работах 😊
    
**[общий комментарий]** Не стоит забывать про оформление своей работы, это важный такой же важный этап. Рекомендую «не стесняться» использовать ячейки типа **markdown** там где нам необходимо отобразить свои размышления, в том числе с использованием дополнительной стилизации.
    
https://sqlbak.com/blog/jupyter-notebook-markdown-cheatsheet
    
    
</div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.